lib/layout/force-directed-layout.js

```diff
@@ -0,0 +1,202 @@
+//force strength
+var forceStr = 1.0/3.0;
+//stretching force
+var stretchForce  = 0.70;
+var vertices;
+var mGraph;
+
+
+//Based partially on https://code.google.com/p/jung/source/browse/branches/
+//guava/jung/jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/
+//layout/SpringLayout.java?r=19
+
+function layOutGraph(graph) {
+	mGraph = graph;
+	vertices = graph.get('vertices');
+	var edgeList = graph.get('edges');
+	var runner = true;
+	while(runner){
+		change = false;
+		for(var i=0;i<edgeList.length;i++){
+			springForce(edgeList[i]);
+		}
+		if(change === false){
+			runner = false;
+		}
+	}
+	//defaulting all ports that do not have a position
+	for(var j=0;j < vertices.length; j++){
+		var inputs = vertices[j].get('inputs');
+		var outputs = vertices[j].get('outputs');
+		
+		for(var h =0; h < inputs.length; h++){
+			
+			if(inputs[h].get('position') === null){
+				inputs[h].set('position', new cardinalPortPosition('E', 50));
+			}
+			
+		}
+		for(h =0;h < outputs.length; h++){
+			if(outputs[h].get('position') === null){
+				outputs[h].set('position', new cardinalPortPosition('W', 50));
+			}
+		}
+	}
+}
+
+function springForce(edge){
+	//vertex to
+	var vertexTo = mGraph.getVertexById(edge.get('to').get('ownerVertexId'));
+	//vertex from
+	var vertexFrom = mGraph.getVertexById(edge.get('from').get('ownerVertexId'));
+	//x distance between them
+	var xLength = edgeXLength(edge);
+	//y distance between them
+	var yLength = edgeYLength(edge);
+	//straight line edge length
+	var length = edgeLength(edge);
+	//checks for both pined
+	if(vertexTo.get('isPinned') ===true && vertexFrom.get('isPinned')===true){
+		return;
+	}
+	
+	//hacky solution for length being 0
+	if(length === 0){
+		length = 0.99;
+	}
+	//repulsion force calculation
+	var force = forceStr * (iLength - length) / length;
+	var pwrTo = vertexTo.get('inputs').length + vertexTo.get('outputs').length;
+	var pwrFr = vertexFrom.get('inputs').length + vertexFrom.get('outputs').length;
+	force = force * Math.pow(stretchForce, (pwrTo + pwrFr - 4));
+	
+	//repulsion change in x
+	var repulsionX = force * xLength;
+	//repulsion change in y
+	var repulsionY = force * yLength;
+	//attraction x
+	var attractionX;
+	//attraction y
+	var attractionY;
+	//math for min length
+	if(length < 1){
+		attractionX = Math.random()*10+1;
+		attractionY = Math.random()*10+1;
+	} else {
+		attractionX = xLength/(length*length);
+		attractionY = yLength/(length*length);
+	}
+	var len = attractionX * attractionX + attractionY * attractionY;
+	//ensures length is positive
+	if(len>0){
+		len = Math.sqrt(len);
+		attractionX = attractionX / len;
+		attractionY = attractionY / len;
+	}
+	//repulsion minus attraction
+	repulsionX = (repulsionX - attractionX);
+	repulsionY = repulsionY - attractionY;
+	var lim = 0.1;
+	
+	if(vertexTo.get('isPinned') ===false && vertexFrom.get('isPinned')===false){
+		vertexTo.set('position', {
+								'x': (vertexTo.get('position').x+repulsionX),
```

Comment:
That is precisely why the use of the tab character is discouraged: because the display is inconsistent between editors. You may have set your editor to make it look fine, but to me it looks absurd.